staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,627 @@
+	// if not then fork an invocation of the EmptyHandler.
+	if qs.NumPacketsEnqueued == 0 && qs.GetRequestsExecuting() == 0 {
+		// fork an invocation of the EmptyHandler.
+		go func() {
+			eh.HandleEmpty()
+		}()
+	}
+	qs.Quiescent = true
+}
+
+// Wait in the happy case, shuffle shards the given request into
+// a queue and eventually dispatches the request from that queue.
+// Dispatching means to return with `quiescent==false` and
+// `execute==true`.  In one unhappy case the request is
+// immediately rebuffed with `quiescent==true` (which tells the
+// filter that there has been a timing splinter and the filter
+// re-calcuates the priority level to use); in all other cases
+// `quiescent` will be returned `false` (even if the system is
+// quiescent by then).  In the non-quiescent unhappy cases the
+// request is eventually rejected, which means to return with
+// `execute=false`.  In the happy case the caller is required to
+// invoke the returned `afterExecution` after the request is done
+// executing.  The hash value and hand size are used to do the
+// shuffle sharding.
+func (qs *QueueSetImpl) Wait(hashValue uint64, handSize int32) (quiescent, execute bool, afterExecution func()) {
+	// TODO(aaron-prindle) verify what should/shouldn't be locked!!!!
+	// TODO(aaron-prindle) collapse all of FQ into one layer/lock (vs 3)
+	//   currently able to collapse to 1 impl layer and 2 locks...
+
+	qs.LockAndSyncTime()
+	// TODO(aaron-prindle) verify and test quiescent
+	// A call to Wait while the system is quiescent will be rebuffed by
+	// returning `quiescent=true`.
+	if qs.Quiescent {
+		return true, false, func() {}
+	}
+
+	// ========================================================================
+	// Step 1:
+	// 1) Start with shuffle sharding, to pick a queue.
+	// 2) Reject old requests that have been waiting too long
+	// 3) Reject current request if there is not enough concurrency shares and
+	// we are at max queue length
+	// 4) If not rejected, create a packet and enqueue
+	pkt := qs.TimeoutOldRequestsAndRejectOrEnqueue(hashValue, handSize)
+	// pkt == nil means that the request was rejected - no remaining
+	// concurrency shares and at max queue length already
+	if pkt == nil {
+		return false, false, func() {}
+	}
+	// ========================================================================
+
+	// ------------------------------------------------------------------------
+	// Step 2:
+	// 1) The next step is to invoke the method that dequeues as much as possible.
+
+	// This method runs a loop, as long as there
+	// are non-empty queues and the number currently executing is less than the
+	// assured concurrency value.  The body of the loop uses the fair queuing
+	// technique to pick a queue, dequeue the request at the head of that
+	// queue, increment the count of the number executing, and send `{true,
+	// handleCompletion(that dequeued request)}` to the request's channel.
+	qs.DequeueWithChannelAsMuchAsPossible()
```

Comment:
every request will be trapping in a loop in this method. considering larger scale cluster w/ 10k+ qps throughput, there will be unnecessary performance loss? how about running this in background as a daemon scheduler? @MikeSpreitzer for thoughts?

Reply:
@yue9944882 can you please elaborate on what you mean by "every request will be trapping in a loop in this method"?  As far as I can see, this call will spend O(N) CPU time where N is the number of requests that get dequeued.  That's as good as we can expect, and this time has to be spent somewhere, and this is the more direct place to put it.

Reply:
`qs.DequeueWithChannelAsMuchAsPossible()` is intended to do only the work that can be done immediately without waiting for any more arrivals or completions.

Reply:
There should be two calls to DequeueWithChannelAsMuchAsPossible in the source code, one for every arrival and one for every completion.  Thus, the average number of requests dequeued per call to DequeueWithChannelAsMuchAsPossible will be no more than 0.5.

Reply:
Is this suggesting that DequeueWithChannelAsMuchAsPossible also be invoked in afterExecution?

Reply:
updated